File: xpose-backend/src/main/java/com/crimereport/xpose/dto/CrimeReportDetail.java
package com.crimereport.xpose.dto;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

public class CrimeReportDetail {
    private String reportId;
    private String crimeType;
    private Integer crimeTypeId;
    private Integer categoryId;
    private String categoryName;
    private String originalDescription;
    private String processedDescription;
    private String address;
    private String city;
    private String state;
    private String policeStation;
    private String status;
    private String urgency;
    private LocalDateTime submittedAt;
    private Double confidenceScore;
    private Double spamScore;
    private Map<String, Object> toxicityScores;
    private Map<String, Object> shapExplanation;
    private String reportQuality;

    private String blockchainHash;
    private LocalDateTime blockchainTimestamp;
    private String blockchainTxId;
    private String rawBlockchainData;

    private Double latitude;
    private Double longitude;
    private Long assignedOfficerId;
    private String assignedOfficerName;

    private String adminStatus;
    private String policeStatus;
    private String policeFeedback;
    private List<String> policeActionProof;
    private LocalDateTime actionTakenAt;
    private Long actionTakenBy;
    private String actionTakenByName;
    private Long reviewedById;
    private String reviewedByName;

    public String getReportId() {
        return reportId;
    }

    public String getCrimeType() {
        return crimeType;
    }

    public Integer getCrimeTypeId() {
        return crimeTypeId;
    }

    public Integer getCategoryId() {
        return categoryId;
    }

    public String getCategoryName() {
        return categoryName;
    }

    public String getOriginalDescription() {
        return originalDescription;
    }

    public String getProcessedDescription() {
        return processedDescription;
    }

    public String getAddress() {
        return address;
    }

    public String getCity() {
        return city;
    }

    public String getState() {
        return state;
    }

    public String getPoliceStation() {
        return policeStation;
    }

    public String getStatus() {
        return status;
    }

    public String getUrgency() {
        return urgency;
    }

    public LocalDateTime getSubmittedAt() {
        return submittedAt;
    }

    public Double getConfidenceScore() {
        return confidenceScore;
    }

    public Double getSpamScore() {
        return spamScore;
    }

    public Map<String, Object> getToxicityScores() {
        return toxicityScores;
    }

    public Map<String, Object> getShapExplanation() {
        return shapExplanation;
    }

    public String getReportQuality() {
        return reportQuality;
    }

    public String getBlockchainHash() {
        return blockchainHash;
    }

    public LocalDateTime getBlockchainTimestamp() {
        return blockchainTimestamp;
    }

    public String getBlockchainTxId() {
        return blockchainTxId;
    }

    public String getRawBlockchainData() {
        return rawBlockchainData;
    }

    public void setReportId(String reportId) {
        this.reportId = reportId;
    }

    public void setCrimeType(String crimeType) {
        this.crimeType = crimeType;
    }

    public void setCrimeTypeId(Integer crimeTypeId) {
        this.crimeTypeId = crimeTypeId;
    }

    public void setCategoryId(Integer categoryId) {
        this.categoryId = categoryId;
    }

    public void setCategoryName(String categoryName) {
        this.categoryName = categoryName;
    }

    public void setOriginalDescription(String originalDescription) {
        this.originalDescription = originalDescription;
    }

    public void setProcessedDescription(String processedDescription) {
        this.processedDescription = processedDescription;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public void setCity(String city) {
        this.city = city;
    }

    public void setState(String state) {
        this.state = state;
    }

    public void setPoliceStation(String policeStation) {
        this.policeStation = policeStation;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public void setUrgency(String urgency) {
        this.urgency = urgency;
    }

    public void setSubmittedAt(LocalDateTime submittedAt) {
        this.submittedAt = submittedAt;
    }

    public void setConfidenceScore(Double confidenceScore) {
        this.confidenceScore = confidenceScore;
    }

    public void setSpamScore(Double spamScore) {
        this.spamScore = spamScore;
    }

    public void setToxicityScores(Map<String, Object> toxicityScores) {
        this.toxicityScores = toxicityScores;
    }

    public void setShapExplanation(Map<String, Object> shapExplanation) {
        this.shapExplanation = shapExplanation;
    }

    public void setReportQuality(String reportQuality) {
        this.reportQuality = reportQuality;
    }

    public void setBlockchainHash(String blockchainHash) {
        this.blockchainHash = blockchainHash;
    }

    public void setBlockchainTimestamp(LocalDateTime blockchainTimestamp) {
        this.blockchainTimestamp = blockchainTimestamp;
    }

    public void setBlockchainTxId(String blockchainTxId) {
        this.blockchainTxId = blockchainTxId;
    }

    public void setRawBlockchainData(String rawBlockchainData) {
        this.rawBlockchainData = rawBlockchainData;
    }

    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }

    public Long getAssignedOfficerId() {
        return assignedOfficerId;
    }

    public void setAssignedOfficerId(Long assignedOfficerId) {
        this.assignedOfficerId = assignedOfficerId;
    }

    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }

    public String getAdminStatus() {
        return adminStatus;
    }

    public void setAdminStatus(String adminStatus) {
        this.adminStatus = adminStatus;
    }

    public String getPoliceStatus() {
        return policeStatus;
    }

    public void setPoliceStatus(String policeStatus) {
        this.policeStatus = policeStatus;
    }

    public String getPoliceFeedback() {
        return policeFeedback;
    }

    public void setPoliceFeedback(String policeFeedback) {
        this.policeFeedback = policeFeedback;
    }

    public List<String> getPoliceActionProof() {
        return policeActionProof;
    }

    public void setPoliceActionProof(List<String> policeActionProof) {
        this.policeActionProof = policeActionProof;
    }

    public LocalDateTime getActionTakenAt() {
        return actionTakenAt;
    }

    public void setActionTakenAt(LocalDateTime actionTakenAt) {
        this.actionTakenAt = actionTakenAt;
    }

    public Long getActionTakenBy() {
        return actionTakenBy;
    }

    public void setActionTakenBy(Long actionTakenBy) {
        this.actionTakenBy = actionTakenBy;
    }

    public String getAssignedOfficerName() {
        return assignedOfficerName;
    }

    public void setAssignedOfficerName(String assignedOfficerName) {
        this.assignedOfficerName = assignedOfficerName;
    }

    public String getActionTakenByName() {
        return actionTakenByName;
    }

    public void setActionTakenByName(String actionTakenByName) {
        this.actionTakenByName = actionTakenByName;
    }

    public String getReviewedByName() {
        return reviewedByName;
    }

    public void setReviewedByName(String reviewedByName) {
        this.reviewedByName = reviewedByName;
    }

    public Long getReviewedById() {
        return reviewedById;
    }

    public void setReviewedById(Long reviewedById) {
        this.reviewedById = reviewedById;
    }

}

File: xpose_app/lib/models/crime_report_detail.dart
import 'package:intl/intl.dart';
import 'dart:convert';

class CrimeReportDetail {
  final String reportId;
  final String crimeType;
  final int? crimeTypeId;
  final int? categoryId;
  final String categoryName;
  final String originalDescription;
  final String? processedDescription;
  final String address;
  final String city;
  final String state;
  final String country;
  final String policeStation;
  final String status;
  final String? urgency;
  final DateTime submittedAt;
  final double? confidenceScore;
  final double? spamScore;
  final Map<String, dynamic>? toxicityScores;
  final Map<String, dynamic>? shapExplanation;
  final String? reportQuality;
  final String? blockchainHash;
  final DateTime? blockchainTimestamp;
  final String? blockchainTxId;
  final double? latitude;
  final double? longitude;
  final int? assignedOfficerId;
  final String? assignedOfficerName;
  final String? rawBlockchainData;
  final String adminStatus;
  final String policeStatus;
  final String? policeFeedback;
  final List<String>? policeActionProof;
  final DateTime? actionTakenAt;
  final int? actionTakenBy;
  final String? actionTakenByName;
  final String? languageDetected;
  final String? translatedDescription;
  final String? rejectionReason;
  final DateTime? reviewedAt;
  final int? reviewedById;
  final String? reviewedByName;

  CrimeReportDetail({
    required this.reportId,
    required this.crimeType,
    this.crimeTypeId,
    this.categoryId,
    required this.categoryName,
    required this.originalDescription,
    this.processedDescription,
    required this.address,
    required this.city,
    required this.state,
    required this.country,
    required this.policeStation,
    required this.status,
    this.urgency,
    required this.submittedAt,
    this.confidenceScore,
    this.spamScore,
    this.toxicityScores,
    this.shapExplanation,
    this.reportQuality,
    this.blockchainHash,
    this.blockchainTimestamp,
    this.blockchainTxId,
    this.latitude,
    this.longitude,
    this.assignedOfficerId,
    this.assignedOfficerName,
    this.rawBlockchainData,
    required this.adminStatus,
    required this.policeStatus,
    this.policeFeedback,
    this.policeActionProof,
    this.actionTakenAt,
    this.actionTakenBy,
    this.actionTakenByName,
    this.languageDetected,
    this.translatedDescription,
    this.rejectionReason,
    this.reviewedAt,
    this.reviewedById,
    this.reviewedByName,
  });

  factory CrimeReportDetail.fromJson(Map<String, dynamic> data) {
    return CrimeReportDetail(
      reportId: data['reportId'] ?? data['id'] ?? '',
      crimeType: data['crimeType'] ?? 'Unknown',
      crimeTypeId: data['crimeTypeId'] is int
          ? data['crimeTypeId']
          : (data['crimeTypeId'] as num?)?.toInt(),
      categoryId: data['categoryId'] is int
          ? data['categoryId']
          : (data['categoryId'] as num?)?.toInt(),
      categoryName: data['categoryName'] ?? 'Unknown',
      originalDescription: data['originalDescription'] ?? '',
      processedDescription:
      data['processedDescription'] ?? data['readabilityEnhancedDescription'],
      address: data['address'] ?? '',
      city: data['city'] ?? '',
      state: data['state'] ?? '',
      country: data['country'] ?? '',
      policeStation: data['policeStation'] ?? '',
      status: data['status'] ?? 'UNKNOWN',
      urgency: data['urgency'] ?? data['urgencyLevel'],
      submittedAt:
      DateTime.parse(data['submittedAt'] ?? DateTime.now().toIso8601String()),
      confidenceScore: data['confidenceScore']?.toDouble(),
      spamScore: data['spamScore']?.toDouble(),
      toxicityScores: data['toxicityScores'] is String
          ? (data['toxicityScores'] != null && data['toxicityScores'].isNotEmpty
          ? Map<String, dynamic>.from(json.decode(data['toxicityScores']))
          : null)
          : (data['toxicityScores'] != null
          ? Map<String, dynamic>.from(data['toxicityScores'])
          : null),
      shapExplanation: data['shapExplanation'] is String
          ? (data['shapExplanation'] != null && data['shapExplanation'].isNotEmpty
          ? Map<String, dynamic>.from(json.decode(data['shapExplanation']))
          : null)
          : (data['shapExplanation'] != null
          ? Map<String, dynamic>.from(data['shapExplanation'])
          : null),
      reportQuality: data['reportQuality'],
      blockchainHash: data['blockchainHash'],
      blockchainTimestamp: data['blockchainTimestamp'] != null
          ? DateTime.parse(data['blockchainTimestamp'])
          : null,
      blockchainTxId: data['blockchainTxId'],
      latitude: data['latitude']?.toDouble(),
      longitude: data['longitude']?.toDouble(),

      assignedOfficerId: data['assignedOfficerId'] is int
          ? data['assignedOfficerId']
          : (data['assignedOfficerId'] as num?)?.toInt(),
      assignedOfficerName: data['assignedOfficerName'],

      rawBlockchainData: data['rawBlockchainData'],
      adminStatus: data['adminStatus'] ?? 'PENDING',
      policeStatus: data['policeStatus'] ?? 'NOT_VIEWED',
      policeFeedback: data['policeFeedback'],
      policeActionProof: data['policeActionProof'] is String
          ? (data['policeActionProof'] != null &&
          data['policeActionProof'].isNotEmpty
          ? List<String>.from(json.decode(data['policeActionProof']))
          : null)
          : (data['policeActionProof'] != null
          ? List<String>.from(data['policeActionProof'])
          : null),
      actionTakenAt:
      data['actionTakenAt'] != null ? DateTime.parse(data['actionTakenAt']) : null,

      actionTakenBy: data['actionTakenBy'] is int
          ? data['actionTakenBy']
          : (data['actionTakenBy'] as num?)?.toInt(),
      actionTakenByName: data['actionTakenByName'],

      languageDetected: data['languageDetected'],
      translatedDescription: data['translatedDescription'],
      rejectionReason: data['rejectionReason'],
      reviewedAt:
      data['reviewedAt'] != null ? DateTime.parse(data['reviewedAt']) : null,

      reviewedById: data['reviewedById'] is int
          ? data['reviewedById']
          : (data['reviewedById'] as num?)?.toInt(),
      reviewedByName: data['reviewedByName'],
    );
  }


  Map<String, dynamic> toJson() {
    return {
      'reportId': reportId,
      'crimeType': crimeType,
      'crimeTypeId': crimeTypeId,
      'categoryId': categoryId,
      'categoryName': categoryName,
      'originalDescription': originalDescription,
      'processedDescription': processedDescription,
      'address': address,
      'city': city,
      'state': state,
      'country': country,
      'policeStation': policeStation,
      'status': status,
      'urgency': urgency,
      'submittedAt': submittedAt.toIso8601String(),
      'confidenceScore': confidenceScore,
      'spamScore': spamScore,
      'toxicityScores': toxicityScores,
      'shapExplanation': shapExplanation,
      'reportQuality': reportQuality,
      'blockchainHash': blockchainHash,
      'blockchainTimestamp': blockchainTimestamp?.toIso8601String(),
      'blockchainTxId': blockchainTxId,
      'latitude': latitude,
      'longitude': longitude,
      'assignedOfficerId': assignedOfficerId,
      'rawBlockchainData': rawBlockchainData,
      'adminStatus': adminStatus,
      'policeStatus': policeStatus,
      'policeFeedback': policeFeedback,
      'policeActionProof': policeActionProof,
      'actionTakenAt': actionTakenAt?.toIso8601String(),
      'actionTakenBy': actionTakenBy,
      'languageDetected': languageDetected,
      'translatedDescription': translatedDescription,
      'rejectionReason': rejectionReason,
      'reviewedAt': reviewedAt?.toIso8601String(),
      'reviewedById': reviewedById,
    };
  }
}